curvlinops/trace/hutchinson.py

```diff
@@ -34,49 +37,44 @@ class HutchinsonTraceEstimator:
+            vector products is greater than the dimension of the linear operator
+            (because then you can evaluate the true trace directly at the same cost).
+
     Example:
-        >>> from numpy import trace, mean, round
+        >>> from numpy import trace, mean
         >>> from numpy.random import rand, seed
         >>> seed(0) # make deterministic
-        >>> A = rand(10, 10)
+        >>> A = rand(50, 50)
         >>> tr_A = trace(A) # exact trace as reference
-        >>> estimator = HutchinsonTraceEstimator(A)
         >>> # one- and multi-sample approximations
-        >>> tr_A_low_precision = estimator.sample()
-        >>> tr_A_high_precision = mean([estimator.sample() for _ in range(1_000)])
-        >>> assert abs(tr_A - tr_A_low_precision) > abs(tr_A - tr_A_high_precision)
+        >>> tr_A_low_precision = hutchinson_trace(A, num_matvecs=1)
+        >>> tr_A_high_precision = hutchinson_trace(A, num_matvecs=40)
+        >>> # compute the relative errors
+        >>> rel_error_low_precision = abs(tr_A - tr_A_low_precision) / abs(tr_A)
+        >>> rel_error_high_precision = abs(tr_A - tr_A_high_precision) / abs(tr_A)
+        >>> assert rel_error_low_precision > rel_error_high_precision
         >>> round(tr_A, 4), round(tr_A_low_precision, 4), round(tr_A_high_precision, 4)
-        (4.4575, 6.6796, 4.3886)
+        (25.7342, 59.7307, 20.033)
     """
-
-    def __init__(self, A: LinearOperator):
-        """Store the linear operator whose trace will be estimated.
-
-        Args:
-            A: Linear square-shaped operator whose trace will be estimated.
-
-        Raises:
-            ValueError: If the operator is not square.
-        """
-        if len(A.shape) != 2 or A.shape[0] != A.shape[1]:
-            raise ValueError(f"A must be square. Got shape {A.shape}.")
-        self._A = A
-
-    def sample(self, distribution: str = "rademacher") -> float:
-        """Draw a sample from the trace estimator.
-
-        Multiple samples can be combined into a more accurate trace estimation via
-        averaging.
-
-        Args:
-            distribution: Distribution of the vector along which the linear operator
-                will be evaluated. Either ``'rademacher'`` or ``'normal'``.
-                Default is ``'rademacher'``.
-
-        Returns:
-            Sample from the trace estimator.
-        """
-        dim = self._A.shape[1]
-        v = random_vector(dim, distribution)
-        Av = self._A @ v
-        return dot(v, Av)
+    if len(A.shape) != 2 or A.shape[0] != A.shape[1]:
+        raise ValueError(f"A must be square. Got shape {A.shape}.")
+    dim = A.shape[1]
+    if num_matvecs >= dim:
+        raise ValueError(
+            f"num_matvecs ({num_matvecs}) must be less than A's size ({dim})."
+        )
+    G = column_stack([random_vector(dim, distribution) for _ in range(num_matvecs)])
+
+    return einsum("ij,ij", G, A @ G) / num_matvecs
```

Comment:
This can be rewritten as a single einsum operation, without first computing `A @ G`.

Reply:
See comment above.